### ISSUE TITLE
[BUG] unable to find non-hoisted dependencies

### ISSUE THREAD

**Issue by ljharb** [vltpkg/vltpkg#267]
- https://github.com/ljharb/get-dep-tree/actions/runs/12190472158/job/34007719903 can't find `set-blocking`
 - https://github.com/es-shims/Symbol.prototype.description/actions/runs/12280032894/job/34265490202#step:12:31 can't find 2 different packages, depending on which run it is
 - https://github.com/inspect-js/typed-array-byte-offset/actions/runs/12287350889/job/34289236279
 - https://github.com/ljharb/list-exports/actions/runs/12288001759/job/34291041640

I'm pretty certain that in all of these cases, the missing dep is indeed explicitly listed in the package.json of the consuming project, and it works fine in an npm-installed environment as well.

**Comment by ljharb** [comment]
One of them comes in via npm, but other than that i cant come up with anything notable.

i do note one of the packages has a “main” set to false - that’s always been valid, but that package has exposed a bug in parcel’s resolver around that (it expects only a string) so maybe that applies here too.

**Comment by isaacs** [comment]
@ljharb This is really weird.

In the latest main of `ljharm/get-dep-tree`, when I run `vlt i` and then `node test/index.js`, it runs fine.

But, when I run `tape test/index.js`, it fails with this:

```
Error: Cannot find module 'cacache'
Require stack:
- /Users/isaacs/Developer/vltpkg/get-dep-tree/node_modules/.vlt/··@npmcli§metavuln-calculator@5.0.1/node_modules/@npmcli/metavuln-calculator/lib/index.js
- {etc...}
```

If I do this:

```bash
$ cat > ./node_modules/.vlt/··@npmcli§metavuln-calculator@5.0.1/node_modules/@npmcli/metavuln-calculator/lib/test-cacache-load.js <<<TEST
console.log(require.resolve('cacache'), require('cacache'))
TEST
$ node ./node_modules/.vlt/··@npmcli§metavuln-calculator@5.0.1/node_modules/@npmcli/metavuln-calculator/lib/test-cacache-load.js

/Users/isaacs/Developer/vltpkg/get-dep-tree/node_modules/.vlt/··cacache@17.1.4/node_modules/cacache/lib/index.js {
  index: {
    compact: [AsyncFunction: compact],
    insert: [AsyncFunction: insert]
  },
  ls: [AsyncFunction: ls] { stream: [Function: lsStream] },
... etc
```

So it's clear that it _can_ resolve from that folder just fine.

When I put a file at `test/cacache-load.js` containing this:

```js
require('.vlt/··@npmcli§metavuln-calculator@5.0.1/node_modules/@npmcli/metavuln-calculator/lib/test-cacache-load.js')
```

Then it works fine when run with either node or tape.

But if I change that to be the `index.js` file, then it works when run with `node test/cacache-load.js` but not `tape test/cacache-load.js`.

Does tape do anything weird to the require paths or something?

**Comment by ljharb** [comment]
No, but it does use the `resolve` package to resolve the module name, and then `require`s that. Does that suggest that vlt breaks `resolve`?

**Comment by isaacs** [comment]
I don't see how it could be, because tape is loading the correct file. Then that file can load the `@npmcli/metavuln-calculator` module just fine, but when mvc tries to load cacache, it can't find it, even though it's clearly discoverable from that path.

I'm confused because it _only_ fails when loaded via tape, and not when loaded via node, and also it _doesn't_ fail with tape when I load a module with just that one require() call.

**Comment by isaacs** [comment]
Ah, the difference seems to be that it loads pacote, and then cacache, and that fails in tape, but not in node directly.

**Comment by isaacs** [comment]
Adding `cacache` as a devDep on the root project makes it work.

Is tape somehow re-using the same require function object across multiple modules, or somehow passing it into a module constructor or something?

**Comment by ljharb** [comment]
the only thing the tape binary does differently here is to use its `require('./import-or-require')` module to evaluate all the files passed in (using `glob`), and `import-or-require` is using `require` or `import()`.

So I suppose it's using the same `require` for all of those entrypoints, but that wouldn't have any bearing on which `require` node itself is using.

**Comment by ljharb** [comment]
Another fun one; this one passed in node 20 but failed on 22 and w3 https://github.com/ljharb/intl-fallback-symbol/actions/runs/12343963527

**Comment by isaacs** [comment]
As far as I can tell, these are only failing when tape (and possibly nyc) is being used, but the modules in question are definitely resolvable using node's resolution directly.

I know that there were some relevant changes to the loaders infrastructure in 22 and 23, perhaps it's related to that? Short of hoisting everything, I'm not sure what vlt can or should do here.

**Comment by ljharb** [comment]
nyc 10 isn't using loaders at all; it's using require.extensions, and tape doesn't use either of those things.

I'm not sure what vlt can do either, but it's still a bug that needs fixing and should probably stay open? if "hoisting everything" is what npm does then that's probably what vlt has to do, tbh.

**Comment by darcyclarke** [comment]
@ljharb from my testing - the initial failures you had should be resolve afaict. I checked out the first few repos & was able to install/run tests just fine &/or eq. to `npm`.

**Comment by ljharb** [comment]
Indeed, this has since been resolved with https://github.com/ljharb/get-dep-tree/actions/runs/18381461268/job/52368689002

If i find new ones, I'll file new ones. Thanks!
